File: 백준/2096(내려가기).cpp
/*
Name : Sohyeon Yim
Email : brightcattle@gmail.com 
Blog : https://sohyunwriter.tistory.com/
Problem : https://www.acmicpc.net/problem/2096
*/

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
	int n;
	cin >> n;
	int minSum, maxSum;
	int a[3];
	int mina[2][3] = { 0, };
	int maxa[2][3] = { 0, };;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < 3; j++) {
			scanf("%d", &a[j]);
		}		
		mina[1][0] = min(a[0] + mina[0][0], a[0] + mina[0][1]);
		maxa[1][0] = max(a[0] + maxa[0][0], a[0] + maxa[0][1]);
		mina[1][1] = min(min(mina[0][0]+a[1], mina[0][1] + a[1]), a[1]+mina[0][2]);
		maxa[1][1] = max(max(maxa[0][0] + a[1], maxa[0][1] + a[1]), a[1]+maxa[0][2]);
		mina[1][2] = min(a[2] + mina[0][1], a[2] + mina[0][2]);
		maxa[1][2] = max(a[2] + maxa[0][1], a[2] + maxa[0][2]);
		
		for (int j = 0; j < 3; j++) {
			mina[0][j] = mina[1][j];
			maxa[0][j] = maxa[1][j];
		}

		//printf("min %d %d %d \n", mina[1][0], mina[1][1], mina[1][2]);
		//printf("max %d %d %d \n", maxa[1][0], maxa[1][1], maxa[1][2]);

	}
	minSum = min(min(mina[1][0], mina[1][1]), mina[1][2]);
	maxSum = max(max(maxa[1][0], maxa[1][1]), maxa[1][2]);
	
	printf("%d %d", maxSum, minSum);

	return 0;



}

File: 백준/2096(내려가기).java
/*
Name : Sohyeon Yim
Email : brightcattle@gmail.com 
Blog : https://sohyunwriter.tistory.com/
Problem : https://www.acmicpc.net/problem/2096
*/

import java.util.Scanner;

public class Main {

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
	
		int N = sc.nextInt();
		int[] a = new int[N*3];
		for(int i = 0; i < N*3; i++) {
			a[i] = sc.nextInt();
		}
		
		long[] min = new long[3];
		long[] max = new long[3];
		
		min[0] = a[0]; min[1] = a[1]; min[2] = a[2];
		max[0] = a[0]; max[1] = a[1]; max[2] = a[2];
		
		for(int i = 1; i < N; i++) {
			//System.out.println(min[2] + " " + max[2]);
			long temp1 = min[0]; long temp2 = min[1]; long temp3 = min[2];
			long temp4 = max[0]; long temp5 = max[1]; long temp6 = max[2]; 
			
			min[0] = a[i*3] + Math.min(min[0], min[1]);
			max[0] = a[i*3] + Math.max(max[0], max[1]);
			
			min[2] = a[i*3 + 2] + Math.min(temp2, temp3);
			max[2] = a[i*3 + 2] + Math.max(temp5, temp6);
			
			
			long mint = Math.min(temp1, temp2);
			mint = Math.min(mint, temp3);
			long maxt = Math.max(temp4, temp5);
			maxt = Math.max(maxt, temp6);
			
			min[1] = a[i*3 + 1] + mint;
			max[1] = a[i*3 + 1] + maxt;
			
		}
		
		
		System.out.println(Math.max(max[2], Math.max(max[0], max[1])) + " " + Math.min(min[2], Math.min(min[0], min[1])) );
		
	}

}
